src/basic_hierarchy/common/HierarchyBuilder.java

```diff
@@ -0,0 +1,395 @@
+package basic_hierarchy.common;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import basic_hierarchy.implementation.BasicNode;
+import basic_hierarchy.interfaces.Node;
+import basic_hierarchy.interfaces.Hierarchy;
+
+
+/**
+ * This class exposes methods allowing to build and fix gaps in {@link Hierarchy} objects.
+ */
+public class HierarchyBuilder
+{
+	private HierarchyBuilder()
+	{
+		// Static class -- disallow instantiation.
+		throw new RuntimeException( "Attempted to instantiate a static class: " + getClass().getName() );
+	}
+
+	/**
+	 * Builds a complete hierarchy of groups, while also patching up holes in the original hierarchy by inserting empty groups
+	 * for missing IDs.
+	 * 
+	 * @param root
+	 *            the root group
+	 * @param groups
+	 *            the original collection of groups
+	 * @param fixBreadthGaps
+	 *            whether the hierarchy fixing algorithm should also fix gaps in breadth, not just depth.
+	 * @param useSubtree
+	 *            whether the centroid calculation should also include child groups' instances.
```

Comment:
Add:
when set to {@code true} then all the objects from subnodes are also regarded as belonging to its supernodes.